libs/go/aealite/protocols/dialogue.go

```diff
@@ -0,0 +1,437 @@
+	dialogueStarterAddress  Address
+}
+
+type Dialogue struct {
+	dialogueLabel     DialogueLabel
+	dialogueMessage   AbstractMessage
+	selfAddress       Address
+	outgoingMessages  []AbstractMessage
+	incomingMessages  []AbstractMessage
+	lastMessageId     int
+	orderedMessageIds []int
+	rules             RuleType
+}
+type AbstractMessage struct {
+	dialogueReference [2]string
+	messageId         MessageId
+	target            Target
+	performative      []string
+	message           []byte
+	to                Address
+	sender            Address
+}
+
+// store dialogues by opponent address
+var dialogueStorage = make(map[Address][]Dialogue)
+
+// store dialogue by dialogue label
+var dialogueByDialogueLabel = make(map[DialogueLabel]Dialogue)
+
+func (dialogue Dialogue) update(message AbstractMessage) {
+	if message.sender == "" {
+		message.sender = dialogue.selfAddress
+	}
+	messageExistence := dialogue.isBelongingToADialogue(message)
+	if !messageExistence {
+		fmt.Println("Error: message does not exist to this dialogue")
+	}
+
+	// TODO
+	// validate next message
+	// check if dialogue message is valid
+	dialogue.validateNextMessage(message)
+	dialogue.updateIncomingAndOutgoingMessages(message)
+}
+
+func (dialogue Dialogue) validateNextMessage(message AbstractMessage) (bool, string) {
+	is_basic_validated, msg_basic_validation := dialogue.basicValidation(message)
+	if !is_basic_validated {
+		return false, msg_basic_validation
+	}
+	// TODO
+	// check if custom validation
+	return true, "Message is valid with respect to this dialogue."
+}
+
+func (dialogue Dialogue) basicValidation(message AbstractMessage) (bool, string) {
+	if dialogue.isEmpty() {
+		return dialogue.basicValidationInitialMessage(message)
+	}
+	return dialogue.basicValidationNonInitialMessage(message)
+}
+
+func (dialogue Dialogue) basicValidationInitialMessage(message AbstractMessage) (bool, string) {
+	dialogue_reference := message.dialogueReference
+	message_id := message.messageId
+	// performative := message.performative
+	if dialogue_reference[0] != dialogue.dialogueLabel.dialogueReference[0] {
+		return false, "Invalid dialogue_reference[0]."
+	}
+	if message_id != StartingMessageId {
+		return false, "Invalid message_id."
+	}
+
+	// TODO
+	// validate message target
+	// err := dialogue.validateMessageTarget(message)
+	// if err != nil{
+	// 	return false, err
+	// }
+
+	// TODO
+	// check if performative exists in intial performatives
+	// if dialogue.rules.initial_performatives {
+	// 	return (
+	//         false,
+	//         "Invalid initial performative."
+	//     )
+	// }
+	return true, "The initial message passes basic validation."
+}
+
+func (dialogue Dialogue) basicValidationNonInitialMessage(message AbstractMessage) (bool, string) {
+	dialogue_reference := message.dialogueReference
+	if dialogue_reference[0] != dialogue.dialogueLabel.dialogueReference[0] {
+		return false, "Invalid dialogue_reference[0]."
+	}
+	err := dialogue.validateMessageId(message)
```

Comment:
`validateMessageId` returns a `string`, but a nullable is expected (maybe `error` is a better return type?)

Reply:
line to change:

https://github.com/fetchai/agents-aea/pull/2373/files#diff-170ecc70a35a8331838e8a1c8489c844807589ac0a14491241ded20ff3c767deR251